mrmustard/math/lattice/strategies/vanilla.py

```diff
@@ -33,109 +33,195 @@
+    This simplifies the bounds check when calculating the index of the pivot :math:`k-1_i`,
+    which need to be done only until the index is smaller than the maximum stride.
+
+    see https://quantum-journal.org/papers/q-2020-11-30-366/ and
+    https://arxiv.org/abs/2209.06069 for more details.
 
     Args:
         shape (tuple[int, ...]): shape of the output tensor
-        A (np.ndarray): A matrix of the Fock-Bargmann representation
-        b (np.ndarray): B vector of the Fock-Bargmann representation
+        A (np.ndarray): A matrix of the Bargmann representation
+        b (np.ndarray): b vector of the Bargmann representation
         c (complex): vacuum amplitude
 
     Returns:
         np.ndarray: Fock representation of the Gaussian tensor with shape ``shape``
     """
-    # calculate the strides
-    strides = shape_to_strides(np.array(shape))
+    # numba doesn't like tuples
+    shape_arr = np.array(shape)
+
+    # calculate the strides (e.g. (100,10,1) for shape (10,10,10))
+    strides = np.ones_like(shape_arr)
+    for i in range(len(shape_arr) - 1, 0, -1):
+        strides[i - 1] = strides[i] * shape_arr[i]
 
     # init flat output tensor
-    ret = np.array([0 + 0j] * np.prod(np.array(shape)))
+    G = np.zeros(np.prod(shape_arr), dtype=np.complex128)
 
-    # initialize the indeces.
-    # ``index`` is the index of the flattened output tensor, while
-    # ``index_u_iter`` iterates through the unravelled counterparts of
-    # ``index``.
-    index = 0
-    index_u_iter = np.ndindex(shape)
-    next(index_u_iter)
+    # initialize the n-dim index
+    nd_index = np.ndindex(shape)
 
-    # write vacuum amplitude
-    ret[0] = c
+    # write vacuum amplitude and skip corresponding n-dim index
+    G[0] = c
+    next(nd_index)
 
-    # iterate over the rest of the indices
-    for index_u in index_u_iter:
-        # update index
-        index += 1
+    # Iterate over the indices smaller than max(strides) with pivot bound check.
+    # The check is needed only if the flat index is smaller than the largest stride.
+    # Afterwards it will be safe to get the pivot by subtracting the first (largest) stride.
+    for flat_index in range(1, strides[0]):
+        index = next(nd_index)
+
+        # calculate (flat) pivot
+        for i, s in enumerate(strides):
+            pivot = flat_index - s
+            if pivot >= 0:  # if pivot not outside array
+                break
+
+        # contribution from pivot
+        value_at_index = b[i] * G[pivot]
 
-        # calculate pivot's contribution
-        i, pivot = first_available_pivot(index, strides)
-        value_at_index = b[i] * ret[pivot]
+        # contributions from pivot's lower neighbours
+        # note the first is when j=i which needs a -1 in the sqrt from delta_ij
+        value_at_index += A[i, i] * SQRT[index[i] - 1] * G[pivot - strides[i]]
+        for j in range(i + 1, len(strides)):
```

Comment:
I'm on my phone so I can't check right now but it's strange that i+2 is equal to the length of strides, because i is the first nonzero entry of an index (so it's variable) and strides is fixed.

Reply:
maybe it was just a coincidence in the examples i was checking.

Reply:
This loop is sometimes triggered but the dimension of `A` and `b` has to be larger than 4x4 and 4.